services/mac_manager/src/main/java/com/futurewei/alcor/macmanager/service/implement/MacServiceImpl.java

```diff
@@ -100,13 +94,26 @@ public MacState getMacStateByMacAddress(String macAddress) throws ParameterNullO
     public MacState createMacState(MacState macState) throws ParameterNullOrEmptyException, MacRepositoryTransactionErrorException, MacRangeInvalidException, MacAddressUniquenessViolationException, MacAddressFullException, MacAddressRetryLimitExceedException {
         if (macState == null)
             throw (new ParameterNullOrEmptyException(MacManagerConstant.MAC_EXCEPTION_PARAMETER_NULL_EMPTY));
+
+        String macAddress = macState.getMacAddress();
+        if(!StringUtils.isEmpty(macAddress)){
+            try {
+                MacState dbMacState = macStateRepository.findItem(macAddress);
+                if(dbMacState != null){
+                    throw new MacAddressUniquenessViolationException("mac address is in use!");
+                }
+                macStateRepository.addItem(macState);
+                return macState;
+            } catch (CacheException e) {
+                throw new MacRepositoryTransactionErrorException("create mac address failed :" + e.getMessage());
+            }
+        }
+
         String rangeId = MacManagerConstant.DEFAULT_RANGE;
```

Comment:
I know this is not your code @Gzure Looking back this line of code will impact the performance significantly. As we are do a major refactor, let us take this opportunity to get rid of the "default range" and use a range selector.

In the MacRangeRepository or a new repository, create a table to track the utilization of each range.

Option 1 __Explicit update__: 
Every allocation request always goes to that table to pick up the latest range with least utilization,  and increment the count after an allocation request or decrease the count after a release request. 
This option is easy to implement at a cost of write performance (more DB access!).

Option 2 __In-Memory Counter with Background Monitoring__:
Each Mac Manager Instance has an in-memory counter that tracks the utilization of each range. The counter gets updated by a running monitoring thread in a periodic manner. This thread should run not very frequently, say every 30 minutes or so, and keep the counter always up-to-date. The allocation/release request only needs to read the in-memory counter but no obligation to update it. 

NOTE that when every Mac Manager instance launches or restart, it also needs to initialize its in-memory counter.